### PR TITLE
Use Codecov GitHub actions app to report coverage

### PR DESCRIPTION
**Pull request by matthewfeickert**
To fix the reporting of coverage (I'm not sure what happened that it broke, as it is a bit hard to trace things at the moment &mdash; probably from an old PR of mine not getting rebased before merge) use the [GitHub Actions Codecov](https://github.com/marketplace/actions/codecov) app to upload and report the coverage.

c.f. https://github.com/codecov/codecov-action

This seems to be the more supported solution then using https://github.com/codecov/codecov-python.

### CODE REVIEW

**Comment by gordonwatts** [comment]
I just realized that though the PR passed tests, after I merged, all the tests started failing. I've not had time to debug that, thanks for looking at this, I'll see what is going on. Since this the "supported" way, I suppose I should get it in.

BTW, this is starting to feel a bit complex and brittle. We shouldn't spend so much time on tools. :-)

**Comment by matthewfeickert** [comment]
> I just realized that though the PR passed tests, after I merged, all the tests started failing. I've not had time to debug that, thanks for looking at this, I'll see what is going on. 

@gordonwatts are you referring to this PR (where I had forgotten to swap out a `platform` for `os` and so GitHub refused to start any of the jobs until I gave it a YAML file that made sense)? Or are you referring to why PR #9 failed upon merge? If so, it is because it was looking for [`codecov` as a pip installed dependency](https://github.com/iris-hep/func_adl/runs/327580143#step:7:6) but it had gotten [removed in the same PR](https://github.com/iris-hep/func_adl/pull/9/files#diff-e9f950f17198d3d5e3122a44230a09b9L29). As that PR got opened up over a month ago I'm not sure why that happened there, as the removal of https://github.com/codecov/codecov-python should have happened in _this_ PR. 

The reason the CI is failing in `master` and not in PR #9 is that [Codecov is set to only run on `push` events](https://github.com/iris-hep/func_adl/blob/d0db9f79fc50c0a3142d81f21221cf7459ce7935/.github/workflows/ci.yml#L38) to that it actually shows the coverage difference. As my PR is coming from a fork GitHub Actions classifies them as `pull_request` events and skips running Codecov (the same thing happened with [your commits that you added to the PR](https://github.com/iris-hep/func_adl/pull/9#issuecomment-560048920)).

> BTW, this is starting to feel a bit complex and brittle. We shouldn't spend so much time on tools. :-)

If you mean "tools" as CI, then sure, but all CI systems are brittle to an extent. GitHub Actions isn't perfect, and as we're seeing right now the patterns that @kratsg and I have developed in pyhf's CI that work well there have some roadblocks when they are transferred to systems where all the PRs aren't coming from the main project but from forks. 

However, I can definitely say that this is _much_ better then if we were using Travis (both @kratsg and I have lost days to debugging locally unreproducible bugs that pop up in Travis). This PR is further moving tooling from the user into the CI system which makes the whole system less brittle to externally installed dependencies that can cause breaks like what is happening now.

**TLDR:** This fixes things by using more infrastructure and less dev (read as: you) maintained code. GitHub Actions can be a bit different then what you expect to happen, but be patient for a little bit longer as this is way better then it used to be.

**Comment by gordonwatts** [comment]
Ok - this looks great now!

**Comment by gordonwatts** [comment]
Ok - so I think this is done now. I ended up doing this organically rather than using your code, I'm afraid, @matthewfeickert  - hope that is ok!

**Comment by matthewfeickert** [comment]
> Ok - so I think this is done now. I ended up doing this organically rather than using your code, I'm afraid, @matthewfeickert - hope that is ok!

That's always fine, but I guess I'm not sure what you mean, as my the code introduced by this PR is still in the repo and [still reporting the coverage](https://github.com/iris-hep/func_adl/runs/534706241?check_suite_focus=true#step:7:37). What did you add?

**Comment by gordonwatts** [comment]
Yeah - no idea - this got added, and then I had to fix something up. I got confused somehow. At any rate, it is all working. :-)

**Comment by matthewfeickert** [comment]
> At any rate, it is all working. :-)

cool. :) If that ever stops being the case feel free to tag me to help fix anything. In the "[pull requests are like puppies](https://twitter.com/brettsky/status/958511462300884992)" analogy, I'm more than happy to feed and walk the dog if you will. :slightly_smiling_face:

**Comment by gordonwatts** [comment]
Awesome. Thanks. :-)
